softlayer/vm/softlayer_creator.go

```diff
@@ -120,9 +125,129 @@ func (c SoftLayerCreator) Create(agentID string, stemcell bslcstem.Stemcell, clo
+	if strings.ToUpper(common.GetOSEnvVariable("OS_RELOAD_ENABLED", "TRUE")) == "FALSE" {
+		return c.CreateNewVM(agentID, stemcell, cloudProps, networks, env)
+	}
+
+	err := bslcvmpool.InitVMPoolDB(bslcvmpool.DB_RETRY_TIMEOUT, bslcvmpool.DB_RETRY_INTERVAL, c.logger)
+	if err != nil {
+		return SoftLayerVM{}, bosherr.WrapError(err, "Failed to initialize VM pool DB")
+	}
+
+	if strings.Contains(cloudProps.VmNamePrefix, "-worker") {
+		vm, err := c.CreateNewVM(agentID, stemcell, cloudProps, networks, env)
+		return vm, err
+	}
+
+	db, err := bslcvmpool.OpenDB(bslcvmpool.SQLITE_DB_FILE_PATH)
+	if err != nil {
+		return SoftLayerVM{}, bosherr.WrapError(err, "Opening DB")
+	}
+
+	vmInfoDB := bslcvmpool.NewVMInfoDB(0, "", "f", "", agentID, c.logger, db)
+	defer vmInfoDB.CloseDB()
+
+	err = vmInfoDB.QueryVMInfobyAgentID(bslcvmpool.DB_RETRY_TIMEOUT, bslcvmpool.DB_RETRY_INTERVAL)
+	if err != nil {
+		return SoftLayerVM{}, bosherr.WrapError(err, "Failed to query VM info by given agent ID "+agentID)
+	}
+
+	if vmInfoDB.VmProperties.Id != 0 {
+		c.logger.Info(SOFTLAYER_VM_CREATOR_LOG_TAG, fmt.Sprintf("OS reload on the server id %d with stemcell %d", vmInfoDB.VmProperties.Id, stemcell.ID()))
+
+		vm := NewSoftLayerVM(vmInfoDB.VmProperties.Id, c.softLayerClient, util.GetSshClient(), nil, c.logger)
+
+		bslcommon.TIMEOUT = 24 * time.Hour
+		err = vm.ReloadOS(stemcell)
+		if err != nil {
+			return SoftLayerVM{}, bosherr.WrapError(err, "Failed to reload OS")
+		}
+
+		if cloudProps.EphemeralDiskSize == 0 {
+			err = bslcommon.WaitForVirtualGuestLastCompleteTransaction(c.softLayerClient, vm.ID(), "Service Setup")
+			if err != nil {
+				return SoftLayerVM{}, bosherr.WrapErrorf(err, "Waiting for VirtualGuest `%d` has Service Setup transaction complete", vm.ID())
+			}
+		} else {
+			err = bslcommon.AttachEphemeralDiskToVirtualGuest(c.softLayerClient, vm.ID(), cloudProps.EphemeralDiskSize, c.logger)
+			if err != nil {
+				return SoftLayerVM{}, bosherr.WrapError(err, fmt.Sprintf("Attaching ephemeral disk to VirtualGuest `%d`", vm.ID()))
+			}
+		}
+
+		virtualGuest, err := bslcommon.GetObjectDetailsOnVirtualGuest(c.softLayerClient, vmInfoDB.VmProperties.Id)
+		if err != nil {
+			return SoftLayerVM{}, bosherr.WrapErrorf(err, "Cannot get details from virtual guest with id: %d.", virtualGuest.Id)
+		}
+
+		softlayerFileService := NewSoftlayerFileService(util.GetSshClient(), virtualGuest, c.logger, c.uuidGenerator, c.fs)
+		agentEnvService := c.agentEnvServiceFactory.New(softlayerFileService, strconv.Itoa(virtualGuest.Id))
+
+		agentEnv := CreateAgentUserData(agentID, cloudProps, networks, env, c.agentOptions)
+		if err != nil {
+			return SoftLayerVM{}, bosherr.WrapErrorf(err, "Cannot agent env for virtual guest with id: %d.", virtualGuest.Id)
+		}
+
+		if len(cloudProps.BoshIp) == 0 {
+			// update /etc/hosts file of bosh-init vm
```

Comment:
It's Jian Qiu's code, checked with Jian Qiu, he suggested to keep it. He will decide if removing it in his  PR.